cuda/matrix/hybrid_kernels.cu

```diff
@@ -0,0 +1,436 @@
+__global__ __launch_bounds__(default_block_size) void fill_in_csr(
+    size_type num_rows, size_type max_nnz_per_row, size_type stride,
+    const ValueType *__restrict__ ell_val,
+    const IndexType *__restrict__ ell_col,
+    const ValueType *__restrict__ coo_val,
+    const IndexType *__restrict__ coo_col,
+    const IndexType *__restrict__ coo_offset,
+    IndexType *__restrict__ result_row_ptrs,
+    IndexType *__restrict__ result_col_idxs,
+    ValueType *__restrict__ result_values)
+{
+    const auto tidx = threadIdx.x + blockDim.x * blockIdx.x;
+
+    if (tidx < num_rows) {
+        auto write_to = result_row_ptrs[tidx];
+        for (auto i = 0; i < max_nnz_per_row; i++) {
+            const auto source_idx = tidx + stride * i;
+            if (ell_val[source_idx] != zero<ValueType>()) {
+                result_values[write_to] = ell_val[source_idx];
+                result_col_idxs[write_to] = ell_col[source_idx];
+                write_to++;
+            }
+        }
+        for (auto i = coo_offset[tidx]; i < coo_offset[tidx + 1]; i++) {
+            if (coo_val[i] != zero<ValueType>()) {
+                result_values[write_to] = coo_val[i];
+                result_col_idxs[write_to] = coo_col[i];
+                write_to++;
+            }
+        }
+    }
+}
+
+
+template <typename ValueType1, typename ValueType2>
+__global__ __launch_bounds__(default_block_size) void add(
+    size_type num, ValueType1 *__restrict__ val1,
+    const ValueType2 *__restrict__ val2)
+{
+    const auto tidx = threadIdx.x + blockDim.x * blockIdx.x;
+    if (tidx < num) {
+        val1[tidx] += val2[tidx];
+    }
+}
+
+
+// Copied from ell_kernel.cu
+template <typename ValueType, typename IndexType>
+__global__ __launch_bounds__(default_block_size) void count_ell_nnz_per_row(
+    size_type num_rows, size_type max_nnz_per_row, size_type stride,
+    const ValueType *__restrict__ values, IndexType *__restrict__ result)
+{
+    constexpr auto warp_size = cuda_config::warp_size;
+    const auto tidx = threadIdx.x + blockIdx.x * blockDim.x;
+    const auto row_idx = tidx / warp_size;
+
+    if (row_idx < num_rows) {
+        IndexType part_result{};
+        for (auto i = threadIdx.x % warp_size; i < max_nnz_per_row;
+             i += warp_size) {
+            if (values[stride * i + row_idx] != zero<ValueType>()) {
+                part_result += 1;
+            }
+        }
+
+        auto warp_tile =
+            group::tiled_partition<warp_size>(group::this_thread_block());
+        result[row_idx] = reduce(
+            warp_tile, part_result,
+            [](const size_type &a, const size_type &b) { return a + b; });
+    }
+}
+
+
+// Copied from coo_kernel.cu
+template <typename IndexType>
+__global__
+    __launch_bounds__(default_block_size) void convert_coo_row_idxs_to_ptrs(
+        const IndexType *__restrict__ idxs, size_type num_nonzeros,
+        IndexType *__restrict__ ptrs, size_type length)
```

Comment:
Here the interface already exist, so including the `coo_kernels.hpp` and calling `coo::convert_row_idxs_to_ptrs(...)` should work.